lib/second_factor/auth_manager.rb

```diff
@@ -0,0 +1,201 @@
+  serializable to JSON.
+
+  :redirect_path => relative subfolder-aware path that the user should be
+  redirected to after the action is finished. When this key is omitted, the
+  redirect path is set to the homepage (/).
+
+  After this method is called, the auth manager will send a 403 response with a
+  JSON body. It does that by raising an exception that's then rescued by a
+  `rescue_from` handler. The JSON response contains a challenge nonce which the
+  client/frontend will need to complete the 2fa. More on this later.
+
+3. second_factor_auth_completed!(callback_params):
+  This method corresponds to outcome (2b) above. It's called after the user has
+  successfully completed the 2fa for the 2fa-protected action and the purpose
+  of this method is to actually perform that action.
+
+  The `callback_params` param of this method is the `callback_params` Hash from
+  the return value of the previous method.
+
+If there are permission/security checks that the current user must pass in
+order to perform the 2fa-protected action, it's important to run the checks in
+all of the 3 methods of the action class and raise errors if the user doesn't
+pass the checks.
+
+Rendering a response to the client in the outcomes (1) and (2b) is a task for
+the controller action. The return value of the `run_second_factor!` method,
+which is an instance of `SecondFactor::AuthManagerResult`, can be used to know
+which outcome the auth manager has picked and render a different response based
+on the outcome.
+
+For a real example where the auth manager is used, please refer to:
+
+* `SecondFactor::Actions::GrantAdmin` action class. This is a class that
+  inherits `SecondFactor::Actions::Base` and implements the 3 methods mentioned
+  above.
+
+* `Admin::UsersController#grant_admin` controller action.
+
+=end
+
+class SecondFactor::AuthManager
+  MAX_CHALLENGE_AGE = 5.minutes
+
+  class SecondFactorRequired < StandardError
+    attr_reader :nonce
+
+    def initialize(nonce:)
+      @nonce = nonce
+    end
+  end
+
+  attr_reader :allowed_methods
+
+  def self.find_second_factor_challenge(nonce, secure_session)
+    challenge_json = secure_session["current_second_factor_auth_challenge"]
+    if challenge_json.blank?
+      raise SecondFactor::BadChallenge.new(
+        "second_factor_auth.challenge_not_found",
+        status_code: 404
+      )
+    end
+
+    challenge = JSON.parse(challenge_json).deep_symbolize_keys
+    if challenge[:nonce] != nonce
+      raise SecondFactor::BadChallenge.new(
+        "second_factor_auth.challenge_not_found",
+        status_code: 404
+      )
+    end
+
+    generated_at = challenge[:generated_at]
+    if generated_at < MAX_CHALLENGE_AGE.ago.to_i
+      raise SecondFactor::BadChallenge.new(
+        "second_factor_auth.challenge_expired",
+        status_code: 401
+      )
+    end
+    challenge
+  end
+
+  def initialize(current_user, guardian, action)
```

Comment:
I have the same concerns here, will we ever pass in a guardian that is not created from the current user?